notify/templates.go

```diff
@@ -0,0 +1,240 @@
+package notify
+
+import (
+	"context"
+	"fmt"
+	tmplhtml "html/template"
+	"net/url"
+	"path/filepath"
+	tmpltext "text/template"
+	"text/template/parse"
+
+	"github.com/grafana/alerting/templates"
+	v2 "github.com/prometheus/alertmanager/api/v2"
+	"github.com/prometheus/alertmanager/notify"
+	"github.com/prometheus/alertmanager/template"
+	"github.com/prometheus/common/model"
+)
+
+type TestTemplatesConfigBodyParams struct {
+	// Alerts to use as data when testing the template.
+	Alerts []*PostableAlert `json:"alerts"`
+
+	// Template string to test.
+	Template string `json:"template"`
+
+	// Name of the template file.
+	Name string `json:"name"`
+}
+
+type TestTemplatesResults struct {
+	Results []TestTemplatesResult      `json:"results"`
+	Errors  []TestTemplatesErrorResult `json:"errors,omitempty"`
+}
+
+type TestTemplatesResult struct {
+	// Name of the associated template definition for this result.
+	Name string `json:"name"`
+
+	// Interpolated value of the template.
+	Text string `json:"text"`
+}
+
+type TestTemplatesErrorResult struct {
+	// Name of the associated template for this error. Will be empty if the Kind is "invalid_template".
+	Name string `json:"name,omitempty"`
+
+	// Kind of template error that occurred.
+	Kind TemplateErrorKind `json:"kind"`
+
+	// Error message.
+	Message string `json:"message"`
+}
+
+type TemplateErrorKind string
+
+const (
+	InvalidTemplate TemplateErrorKind = "invalid_template"
+	ExecutionError  TemplateErrorKind = "execution_error"
```

Comment:
Nit: subsequent definitions should inherit the type from the above

```suggestion
	ExecutionError  = "execution_error"
```

Reply:
This change gets caught by the linter as `SA9004: only the first constant in this group has an explicit type (staticcheck)`